### ISSUE TITLE
[Bug]: Nonce Mismatch During Refresh Token Flow with Keycloak (OIDC)

### ISSUE THREAD

**Issue by fengccp** [oauth2-proxy/oauth2-proxy#2876]
### OAuth2-Proxy Version

v7.7.1

### Provider

keycloak-oidc

### Expected Behaviour

When using the refresh token flow, OAuth2-Proxy should be able to refresh the access token without encountering a nonce mismatch.

https://www.keycloak.org/docs/latest/upgrading/#nonce-claim-is-only-added-to-the-id-token

### Current Behaviour

Keycloak does not include a nonce in the ID token when refreshing tokens. Because of this, OAuth2-Proxy throws an error for nonce mismatch when `insecureSkipNonce` is set to `false`

### Steps To Reproduce

1. Set up OAuth2-Proxy with Keycloak as the identity provider.
2. Configure the proxy with insecureSkipNonce: false.
3. Attempt to refresh the token using the refresh token flow.
4. Observe the nonce mismatch error.

### Configuration details or additional information

For now, I've added a mapper in Keycloak called Nonce backwards compatible. This temporarily adds a nonce to the ID token after refreshing, which allows OAuth2-Proxy to refresh tokens successfully.

**Comment by JoelSpeed** [comment]
So the issue specifically here is that we are using the same nonce checking on the initial token, and refreshed tokens, and I'm guessing some providers do provide a nonce, and some do not?

If you set `insecureSkipNonce` this fixes the issue, but also means that the nonce is not being checked for the initial ID token either?

I guess we probably want to skip the check on refresh tokens, per the spec link you added.

Anyone able to look at wrangling up a PR?

**Comment by m4123** [comment]
Yes, if we set `insecureSkipNonce = true`, the issue is not present. But then the nonce check is also disabled for the initial ID token.
Based on https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse, I expect the following behavior:

1.  `insecureSkipNonce  = true`: Do not validate the nonce, neither in the initial ID token, nor in refreshed ones.
2. `insecureSkipNonce  = false`: Validate the nonce in the initial ID token. Validate the nonce in a refreshed ID token only if the nonce claim is present in the refreshed ID token (if present, the nonce of refreshed ID tokens MUST be the same as the one in the initial ID token).

**Comment by kennylam91** [comment]
We had the same issue and opted to add a Nonce backwards compatible mapper in Keycloak so as not to impact overall security. 
We're waiting for the improvement of nonce check behavior.

**Comment by m4123** [comment]
Any update when a fix could be merged? Thanks a lot for your efforts!
